File: java/src/main/java/org/twyszomirski/aoc/solutions/Day_3.java
package org.twyszomirski.aoc.solutions;

import org.twyszomirski.aoc.utils.Pair;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static java.io.IO.println;

public class Day_3 {

    public static void solve() throws IOException {
        var lines = Files.readAllLines(Paths.get("src/main/resources/input_day_3.txt"));
        println("======== Day 3 ===========");
        part_1(lines);
        part_2(lines);
    }

    public static void part_1(List<String> lines) {
        var input = String.join("", lines);
        var mulPattern = Pattern.compile("mul\\((\\d{1,3}),(\\d{1,3})\\)");
        Matcher m = mulPattern.matcher(input);

        var results = new ArrayList<Pair<Integer>>();
        while (m.find()) {
            var group = m.group(2);
            results.add(new Pair<>(Integer.valueOf(m.group(1)), Integer.valueOf(m.group(2))));
        }

        var sum = results.stream().mapToInt(it -> it.left() * it.right()).sum();
        println(sum);

    }

    public static void part_2(List<String> lines) {
        var input = String.join("", lines);
        var dontPattern = Pattern.compile("don't\\(\\)");
        Matcher dontMatcher = dontPattern.matcher(input);

        var doPattern = Pattern.compile("do\\(\\)");
        Matcher doMatcher = doPattern.matcher(input);

        var results = new ArrayList<String>();
        var currentIdx = 0;

        while (currentIdx < input.length()) {
            var nextDont = dontMatcher.find(currentIdx) ? dontMatcher.start() : input.length();
            results.add(input.substring(currentIdx, nextDont));
            currentIdx = doMatcher.find(nextDont) ? doMatcher.start() : input.length();
        }

        part_1(results);
    }

}

File: kotlin/src/main/kotlin/org/twyszomirski/aoc/solutions/Day_3.kt
package org.twyszomirski.aoc.solutions

import java.io.File
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sign


class Day_3 {

    fun solve() {
        val lines = File("src/main/resources/input_day_3.txt").readLines()
        println("======== Day 3 ===========")
        part1(lines)
        part2(lines)
    }

    fun part1(lines: List<String>) {
        val sum = lines.joinToString(separator = "")
            .windowedSequence(12)
            .mapNotNull { window ->
                if (window.startsWith("mul(")) {
                    parse(window.substring(4)).result
                } else null
            }
            .sumOf { it.first * it.second }

        println(sum)
    }

    fun parse(candidate: String): ParseResult {
        val firstNumber = getPostFixedNumber(candidate.substring(0, min(4, candidate.length)), ',')
        if (firstNumber != null) {
            val secondNumber = getPostFixedNumber(candidate.substring(firstNumber.length + 1), ')')
            if (secondNumber != null) {
                return ParseResult(
                    firstNumber.length + secondNumber.length + 2,
                    Pair(firstNumber.toInt(), secondNumber.toInt())
                )
            }
        }
        return ParseResult(0, null)
    }

    fun getPostFixedNumber(input: String, postfix: Char): String? {
        if (!input.contains(postfix)) return null
        return input.substringBefore(postfix).takeIf { it.all { char -> char.isDigit() } }
    }


    fun part2(lines: List<String>) {
        val input = lines.joinToString(separator = "")

        var currentIdx = 0
        val result = mutableListOf<String>()
        var nextDont = input.indexOf("don't()")

        while (currentIdx < input.length) {
            result.add(input.substring(currentIdx, nextDont))
            currentIdx = input.indexOf("do()", nextDont).takeIf { it > 0 } ?: input.length
            nextDont = input.indexOf("don't()", currentIdx).takeIf { it > 0 } ?: input.length
        }
        val trimmed = result.joinToString(separator = "")
        part1(listOf(trimmed))
    }


}

data class ParseResult(val advance: Int, val result: Pair<Int, Int>?)